### ISSUE TITLE
[DR] Reorganize GK I/O functions to allow infrequent writing of phase-space quantities

### ISSUE THREAD

**Issue by manauref** [ammarhakim/gkylzero#483]
Presently the distribution function gets written out every frame. That's because the function `gkyl_gyrokinetic_app_write` which writes the distribution also writes other stuff (mostly configuration-space diagnostics). This is problematic because most applications don't need the distribution at the same frame rate, yet they are still filling up our disk space. See issue #478.

A solution, prototyped in the gk-g0-app_io_reorg branch, is to split the contents of `gkyl_gyrokinetic_app_write` into two other functions
```
void gkyl_gyrokinetic_app_write_phase(gkyl_gyrokinetic_app* app, double tm, int frame);
void gkyl_gyrokinetic_app_write_conf(gkyl_gyrokinetic_app* app, double tm, int frame);
```
which as their name suggest, are responsible for writing phase-space and conf-space diagnostics, respectively. `gkyl_gyrokinetic_app_write` will simply call these two, so that input files that continue to to call `gkyl_gyrokinetic_app_write` are not affected (arguably we should deprecate that at some point). Additionally, users that wish to output phase-space quantities less frequently could call `gkyl_gyrokinetic_app_write_phase` according to a different trigger, like (e.g. in write_data of our input files):
```
void
write_data(struct gkyl_tm_trigger* iot_conf, struct gkyl_tm_trigger* iot_phase, gkyl_gyrokinetic_app* app, double t_curr, bool force_write)
{
  bool trig_now;

  trig_now = gkyl_tm_trigger_check_and_bump(iot_conf, t_curr);
  if (trig_now || force_write) {
    int frame = (!trig_now) && force_write? iot_conf->curr : iot_conf->curr-1;

    gkyl_gyrokinetic_app_write_conf(app, t_curr, frame);

    gkyl_gyrokinetic_app_calc_mom(app);
    gkyl_gyrokinetic_app_write_mom(app, t_curr, frame);
  }

  trig_now = gkyl_tm_trigger_check_and_bump(iot_phase, t_curr);
  if (trig_now || force_write) {
    int frame = (!trig_now) && force_write? iot_phase->curr : iot_phase->curr-1;

    gkyl_gyrokinetic_app_write_phase(app, t_curr, frame);
  }
}
```

Looking at this seems a bit confusing, as mom is a conf-space diagnostic, so not sure if we should really be using different functions to call it (though we could leave the functions in the code for greater fine-graining).

**Comment by JunoRavin** [comment]
How are you thinking about handling the restart capabilities? Do we need to pass two flags for restarts now (the phase space frame number and the configuration space frame number) or are the output frames in phase space when writing out less frequently will still map onto the configuration space frame numbers (so you may write out the distribution function at "frame" 5, 10, 15,... )

**Comment by manauref** [comment]
I'm hoping we still simply provide a single frame number, and the input file computes the corresponding frame number for the distributions.

For example:
Assume a simulation with num_frames=100 and phase_write_freq=0.2 (so we get 20 distributions).

sim 1:
- t_end = 3.0;
- writes 100 frames of moments, and 20 frames of distributions

sim 2 (restart):
- t_end = 6.0;
- run with `-r 100` to restart (so restart_frame=100)
- given `phase_write_freq=0.2` the code will look for frame `phase_write_freq*restart_frame=20` and load the 20th distribution.

What do you think?

**Comment by manauref** [comment]
The above convention is what we had in g2, where the number of frames really meant the number of conf-space frames.
See a demonstration of an input file that can restart as described above here:
https://github.com/ammarhakim/gkylzero/commit/816a869372fa2de25e6167b2d04361ea5e812b13

Another option is to simply make the `-r` value be the distribution function frame we are restarting from. But I prefer not to do this and to stick to the g2 convention instead.

**Comment by Antoinehoff** [comment]
What about keeping the same numbering for both phase and conf and just skip frames with phase? Like in sim 1, conf will go for 0,1,2...,100 and phase for 0,5,10,...,100? It's then easier to link them when we look at the files for postprocessing imo

**Comment by JunoRavin** [comment]
I lean towards @Antoinehoff's solution because I worry about causing headaches if someone changes the I/O frequency restart to restart, or requests some length of time in a restart where they should change the I/O frequency to make the distribution function cadence and configuration space function cadence match up. 

The only other thing I can think of which might work: can we store the distribution function frame as meta-data in configuration space files? Then we could pass the configuration space frame to the restart command and read the meta-data to figure out what distribution function frame to read in.

**Comment by ammarhakim** [comment]
@JunoRavin We can't store distribution function frame in the conf quantities meta files.

Actually, all this represents a confusion which is perhaps understandable: remember the **C driver is specifying the frame numbers**. This means the C driver can specify arbitrary frame numbers, for example, the sequence of primes, or perfect numbers or Fibonacci sequence. Of course, one should avoid such perverse things, but this is not disallowed.

Hence, ensuring that the restart is done properly is also the responsibility of the C driver. The `-r` flag is just something our regression test provide, and not something universal for every C driver. As each C driver can be unique there is no way to make the `-r` flag work in the same work in each driver.

Hence, this means the issue of restarts is not connected to that of frame numbers. If you use  a perverse numbering scheme that it your responsibility to ensure that restarts are consistent.

Finally, I want to point out that universal behavior of the `-r` flag is only possible via Lua driver. It is hence we complete those as soon as we can.

**Comment by ammarhakim** [comment]
One more important point: We must not remove the species index from the species write method. Hence, from the C driver it should continue to be possible to write electrons and ions at different frequencies. Please remember: the C API must be flexible and sufficiently fine-grained. It is the responsibility of the driver (C or Lua or whatever you like) to call these.

**Comment by ammarhakim** [comment]
Please methods of the form:

```void gkyl_gyrokinetic_app_write_species_phase(gkyl_gyrokinetic_app* app, int sidx, double tm, int frame)
```
etc to the system. Then call these in the new `app_write` methods proposed above. 

Please keep in mind the API design principle: think of API flexibility and granularity. We can then add wrappers around those to make life a bit easier. But without a flexible API we can't produce a powerful and extensible code.

**Comment by manauref** [comment]
I'll incorporate above comments:
- keep numbering for phase frames the same as that for conf frames.
- add more similar methods for species.

**Comment by manauref** [comment]
Addressed in PR #484
